Move cmd/controller-manager to k8s.io/controller-manager

cici37 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug

 /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

This PR is to address the concern: https://github.com/kubernetes/kubernetes/pull/90976#issuecomment-635546451
Moved cmd/controller-manager to k8s.io/controller-manager
Moved GenericControllerManagerConfiguration in k8s.io/controller-manager. 
Moved CloudControllerManagerConfiguration, KubeCloudSharedConfiguration and ServiceControllerConfiguration to k8s.io/cloud-provider.


**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

cici37 commented:
cc @cheftako @andrewsykim This pr is for decoupling types between CCM and KCM. And it will also make it ready to move cmd/controller-manager out-tree if we plan to do so.

cici37 commented:
/ok-to-test

fedebongio commented:
/assign @cheftako

cheftako commented on staging/src/k8s.io/cloud-provider/api/types.go:
```diff
@@ -0,0 +1,93 @@
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// GenericControllerManagerConfiguration holds configuration for a generic controller-manager
+type GenericControllerManagerConfiguration struct {
```
We want this shared between the KCM and CCM. I don't think we want to decouple this.

cici37 commented:
Will have the PR ready after new k8s.io/controller-manager repo is in place.

cici37 commented:
cc @cheftako @andrewsykim

cheftako commented on staging/src/k8s.io/controller-manager/app/options/types.go:
```diff
@@ -0,0 +1,99 @@
+	componentbaseconfig "k8s.io/component-base/config"
+)
+
+// ServiceControllerConfiguration contains elements describing ServiceController.
```
Can we put the controller configs in sub-directories. Separating controller concerns from controller-manager concerns seems prudent. It seems fine to have service controller here, while its jsut service controller. However when it becomes, service, cloud, cloud lifecycle, route, x, y and z controllers, I think we will want them each in their own sub directory.

cici37 commented:
/retest

cheftako commented on pkg/controller/service/config/OWNERS:
```diff
@@ -1,14 +0,0 @@
-approvers:
-- api-approvers
```
What about https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/config/v1alpha1/zz_generated. deepcopy.go? Either it should also be deleted or we should retain OWNERS.

cici37 replied:
`https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/config/v1alpha1/zz_generated.deepcopy.go` is deleted together with OWNERS file in this PR.

cheftako replied:
Sorry somehow missed it when I looked last time. Not sure how I missed it.

cheftako commented on staging/src/k8s.io/controller-manager/app/options/v1alpha1/conversion.go:
```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
2020

cici37 commented on staging/src/k8s.io/controller-manager/app/options/types.go:
```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
@cheftako Should here be 2019 or 2020? Thanks

cheftako replied:
Old files should be left with their original Copyright notice. New files are generally generated with the current year.

cici37 commented:
/retest

cici37 commented:
Wait for PR [92399](https://github.com/kubernetes/kubernetes/pull/92399) to be merged first

cici37 commented:
I have combined changes of [92399](https://github.com/kubernetes/kubernetes/pull/92399) here.

cici37 commented:
/retest

cici37 commented:
/retest

cici37 commented:
/retest

cici37 commented:
/retest

cici37 commented:
/retest

cici37 commented:
/retest

cici37 commented:
/retest

cici37 commented:
/retest

cici37 commented:
/retest

andrewsykim commented on cmd/cloud-controller-manager/app/apis/config/types.go:
```diff
@@ -18,8 +18,8 @@ package config
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	kubectrlmgrconfig "k8s.io/kubernetes/pkg/controller/apis/config"
-	serviceconfig "k8s.io/kubernetes/pkg/controller/service/config"
+	config2 "k8s.io/controller-manager/config"
```
controllermanagerconfig or cmconfig

cici37 replied:
Updated

andrewsykim commented on cmd/kube-controller-manager/app/options/options_test.go:
```diff
@@ -24,13 +24,14 @@ import (
 	componentbaseconfig "k8s.io/component-base/config"
 	"k8s.io/component-base/metrics"
-	cmoptions "k8s.io/kubernetes/cmd/controller-manager/app/options"
+	config2 "k8s.io/controller-manager/config"
```
same here, cmconfig or controllermanagerconfig

andrewsykim commented on staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go:
```diff
@@ -18,8 +18,9 @@ package v1alpha1
 
 import (
 	"k8s.io/apimachinery/pkg/conversion"
+	config2 "k8s.io/controller-manager/config"
```
cmconfig or controllermanagerconfig

andrewsykim commented on staging/src/k8s.io/controller-manager/options/cloudprovider.go:
```diff
@@ -18,13 +18,12 @@ package options
 	"github.com/spf13/pflag"
-
-	kubectrlmgrconfig "k8s.io/kubernetes/pkg/controller/apis/config"
+	config2 "k8s.io/controller-manager/config"
```
cmconfig or controllermanagerconfig

andrewsykim commented on staging/src/k8s.io/controller-manager/options/generic.go:
```diff
@@ -23,19 +23,19 @@ import (
 	cliflag "k8s.io/component-base/cli/flag"
 	"k8s.io/component-base/config/options"
-	kubectrlmgrconfig "k8s.io/kubernetes/pkg/controller/apis/config"
+	config2 "k8s.io/controller-manager/config"
```
cmconfig or controllermanagerconfig

andrewsykim commented on staging/src/k8s.io/controller-manager/options/kubecloudshared.go:
```diff
@@ -18,25 +18,24 @@ package options
 	"github.com/spf13/pflag"
-
-	kubectrlmgrconfig "k8s.io/kubernetes/pkg/controller/apis/config"
+	config2 "k8s.io/controller-manager/config"
```
cmconfig or controllermanagerconfig

cici37 commented:
/retest

cici37 commented:
/retest

cici37 commented:
/retest

nikhita commented:
/test pull-publishing-bot-validate

nikhita commented:
publishing-bot job is green :tada:

cici37 commented:
@nikhita Thanksssss so much for helping out on this test :)

cheftako commented on pkg/controller/apis/config/types.go:
```diff
@@ -18,7 +18,8 @@ package config
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	componentbaseconfig "k8s.io/component-base/config"
+	config2 "k8s.io/controller-manager/config"
```
Can we make this cmconfig rather than config2? We may also want to alias service/config as serviceconfig. If we did that then controller-managers config could just be config. As the base config that would seem to make more sense.

cici37 replied:
Updated to use `cmconfig` for /control-manager/config and `serviceconfig` for /service/config

cheftako commented:
/retest

cheftako commented on cmd/cloud-controller-manager/app/options/options_test.go:
```diff
@@ -23,22 +23,21 @@ import (
-	serviceconfig "k8s.io/kubernetes/pkg/controller/service/config"
+	cmconfig "k8s.io/controller-manager/config"
+	cmoptions "k8s.io/controller-manager/options"
+	"k8s.io/controller-manager/service/config"
```
Can we keep this as serviceconfig?

cheftako commented on cmd/cloud-controller-manager/app/apis/config/v1alpha1/defaults.go:
```diff
@@ -21,8 +21,8 @@ import (
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	serviceconfigv1alpha1 "k8s.io/controller-manager/config/v1alpha1"
```
This should probably become cmconfigv1alpha1 or ctrlmgrconfigv1alpha1.

andrewsykim replied:
+1

cheftako commented on staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2018 The Kubernetes Authors.
+Copyright 2020 The Kubernetes Authors.
```
New files should use the current year. I believe old files (including those being moved) should be left with their original year.

cheftako commented on staging/src/k8s.io/controller-manager/service/config/doc.go:
```diff
@@ -0,0 +1,19 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
New file 2020.

andrewsykim reviewed (commented):
LGTM, have mostly the same minor comments as @cheftako

cheftako commented:
/retest

cheftako commented:
/priority important-soon

cici37 commented:
/test pull-kubernetes-e2e-kind

cheftako commented:
/retest

cici37 commented:
/test pull-kubernetes-e2e-kind

cici37 commented:
/test pull-kubernetes-e2e-kind

cici37 commented:
/test pull-kubernetes-e2e-kind

cheftako commented:
/test pull-kubernetes-e2e-kind

cici37 commented:
/test pull-kubernetes-e2e-kind

cici37 commented:
/test pull-kubernetes-e2e-kind

cici37 commented:
/test pull-kubernetes-e2e-kind

cici37 commented:
/retest

cici37 commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

lavalamp commented:
it looks like this was approved before freeze? is that right? Should we put it in the milestone?

cheftako commented:
/milestone 1.19

cheftako commented:
/milestone v1.19

cheftako commented:
> it looks like this was approved before freeze? is that right? Should we put it in the milestone?

Will that be sufficient?

lavalamp commented:
Wait, my bad, it looks like this didn't actually have an approved label? :(

andrewsykim reviewed (commented):
Can we file an exception for this? Seems pretty low risk and high benefit for external cloud providers consuming cmd/cloud-controller-manager today.

lavalamp commented:
We can certainly try! If it's low risk and blocking something important, I
think you can make a case? I didn't realize it was blocking something, or I
would have suggested it.

On Wed, Jul 15, 2020 at 8:56 AM Andrew Sy Kim <notifications@github.com>
wrote:

> *@andrewsykim* commented on this pull request.
>
> Can we file an exception for this? Seems pretty low risk and high benefit
> for external cloud providers consuming cmd/cloud-controller-manager today.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/91014#pullrequestreview-449079852>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE6BFVYORMGB44RK7HB3JTR3XGRNANCNFSM4M6UE5GA>
> .
>

nilo19 commented:
what's blocking it? really need this to get merged.

cici37 commented:
Thanks @cheftako to help on communicating with release manager to see if we could get it to 1.19. Still in progress

cheftako commented:
/lgtm

cheftako commented:
/approve

dims commented:
/assign @smarterclayton @liggitt 

Clayton, Jordan,
Needs root approval

cici37 commented:
/retest

dims commented:
/uncc

cici37 commented:
Gentle reminder :) @smarterclayton @liggitt

cici37 commented:
/retest

cici37 commented:
/retest

cici37 commented:
/test pull-kubernetes-node-e2e

cheftako commented:
/lgtm
/approve

cici37 commented:
/assign @liggitt

cici37 commented:
/retest

cici37 commented:
/test pull-publishing-bot-validate

cici37 commented:
/test pull-publishing-bot-validate

cheftako commented on staging/src/k8s.io/controller-manager/service/config/BUILD:
```diff
@@ -7,7 +7,8 @@ go_library(
         "zz_generated.deepcopy.go",
     ],
-    importpath = "k8s.io/kubernetes/pkg/controller/service/config",
+    importmap = "k8s.io/kubernetes/vendor/k8s.io/controller-manager/service/config",
```
service/config used to have things like an OWNERS file. Is there a reason we seem to have lost that?

cici37 replied:
Thanks for the catch. Updated^

cheftako commented on staging/src/k8s.io/controller-manager/service/config/doc.go:
```diff
@@ -1,7 +1,5 @@
-// +build !ignore_autogenerated
```
This move seems wrong. I think we'll end up with the wrong file history.

cici37 replied:
Thanks. Updated to remove this file history

cici37 commented:
/test pull-publishing-bot-validate

cici37 commented:
/test pull-publishing-bot-validate

cici37 commented:
/test pull-publishing-bot-validate

cici37 commented:
/test pull-publishing-bot-validate

cici37 commented:
/test pull-publishing-bot-validate

cheftako commented on staging/publishing/rules.yaml:
```diff
@@ -31,7 +31,7 @@ rules:
       dir: staging/src/k8s.io/code-generator
     name: release-1.19
-    go: 1.15.2
+    go: 1.15.2.2
```
Can we undo this change?

cici37 replied:
Updated

cheftako commented on staging/publishing/rules.yaml:
```diff
@@ -59,7 +59,7 @@ rules:
       dir: staging/src/k8s.io/apimachinery
     name: release-1.19
-    go: 1.15.2
+    go: 1.15.2.2
```
Ditto

cici37 commented:
/test pull-kubernetes-e2e-kind-ipv6

liggitt commented on cmd/cloud-controller-manager/app/apis/config/types.go:
```diff
@@ -29,10 +29,10 @@ type CloudControllerManagerConfiguration struct {
 	// KubeCloudSharedConfiguration holds configuration for shared related features
 	// both in cloud controller manager and kube-controller manager.
-	KubeCloudShared kubectrlmgrconfig.KubeCloudSharedConfiguration
+	KubeCloudShared cmconfig.KubeCloudSharedConfiguration
```
GenericControllerManagerConfiguration in k8s.io/controller-manager makes sense

KubeCloudSharedConfiguration and ServiceController types should go to k8s.io/cloud-provider, to avoid putting cloud-specific and kubernetes-specific stuff into the generic controller manager repo. KCM can depend on k8s.io/cloud-provider until it stops running cloud loops.

cici37 replied:
Updated the PR to move KubeCloudSharedConfiguration and ServiceController types to k8s.io/cloud-provider. Please have a look when you got time😊

liggitt commented on staging/src/k8s.io/controller-manager/go.mod:
```diff
@@ -5,16 +5,24 @@ module k8s.io/controller-manager
 	k8s.io/apimachinery v0.0.0
 	k8s.io/apiserver v0.0.0
+	k8s.io/client-go v0.0.0
+	k8s.io/cloud-provider v0.0.0
 	k8s.io/component-base v0.0.0
+	k8s.io/klog/v2 v2.2.0
+	k8s.io/kube-controller-manager v0.0.0
```
these two dependencies should not exist... I expected cloud-provider and kube-controller-manager to depend on the generic controller-manager module, not the other way around

cheftako replied:
A little uncomfortable that we may have put too much (interface definition and controllers) into cloud-provider. Topic for another time.

liggitt commented on cmd/cloud-controller-manager/app/apis/config/v1alpha1/defaults.go:
```diff
@@ -47,5 +47,5 @@ func SetDefaults_CloudControllerManagerConfiguration(obj *CloudControllerManager
 	kubectrlmgrconfigv1alpha1.RecommendedDefaultGenericControllerManagerConfiguration(&obj.Generic)
 	// Use the default RecommendedDefaultServiceControllerConfiguration options
-	serviceconfigv1alpha1.RecommendedDefaultServiceControllerConfiguration(&obj.ServiceController)
+	cmconfigv1alpha1.RecommendedDefaultServiceControllerConfiguration(&obj.ServiceController)
```
I expected the defaulting functions for the service controller to accompany the service controller config type and be in `k8s.io/cloud-provider`

cheftako replied:
+1

cici37 replied:
Updated.

liggitt commented on staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go:
```diff
@@ -18,8 +18,10 @@ package v1alpha1
+	cpconfig "k8s.io/cloud-provider/config"
+	serviceconfig "k8s.io/cloud-provider/service/config"
+	cmconfig "k8s.io/controller-manager/config"
 	"k8s.io/kube-controller-manager/config/v1alpha1"
```
the v1alpha1 GenericControllerManagerConfiguration type should move to k8s.io/controller-manager as well so this dependency is dropped

liggitt commented on staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go:
```diff
@@ -18,8 +18,10 @@ package v1alpha1
 
 import (
 	"k8s.io/apimachinery/pkg/conversion"
+	cpconfig "k8s.io/cloud-provider/config"
+	serviceconfig "k8s.io/cloud-provider/service/config"
```
the v1alpha1 KubeCloudSharedConfiguration and ServiceControllerConfiguration types (and the conversions from this file) should move to `k8s.io/cloud-provider` to drop these dependencies

liggitt commented on staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go:
```diff
@@ -18,8 +18,10 @@ package v1alpha1
 
```
this file should be split up so that this repo only contains conversions for GenericControllerManagerConfiguration, and the rest of the conversions are moved to the cloud-provider repo

cici37 commented:
/test pull-publishing-bot-validate

nikhita commented:
```
 F1008 07:46:48.473993     157 main.go:35] Invalid rules file "/home/prow/go/src/k8s.io/kubernetes/staging/publishing/rules.yaml": validation errors:
- repository "kube-controller-manager" cannot depend on "controller-manager" later in the rules file
- repository "kube-controller-manager" cannot depend on "cloud-provider" later in the rules file 
```

the `controller-manager` rules need to be moved above the `kube-controller-manager` rules in rules.yaml for the publishing-bot job to pass

cici37 committed:
Move cmd/controller-manager to k8s.io/controller-manager and cloud specific configs to k8s.io/cloud-provider.

cici37 commented:
@nikhita Thank you so much for the details 😊

cici37 commented:
Brief of this PR:

1. Moved `cmd/controller-manager` to `k8s.io/controller-manager`;
    - GenericControllerManagerConfiguration(from `pkg/controller/apis/config`) to `k8s.io/controller-manager/config`;

2. Moved `pkg/controller/service` to `k8s.io/cloud-provider/service` (ServiceControllerConfiguration);

3. Moved `cmd/cloud-controller-manager/app/apis/config` to `k8s.io/cloud-provider/app/apis/config`
    - CloudControllerManagerConfiguration(from `cmd/cloud-controller-manager/app/apis/config`), KubeCloudSharedConfiguration(from `pkg/controller/apis/config`) and CloudProviderConfiguration(from `pkg/controller/apis/config`) to `k8s.io/cloud-provider/app/apis/config`;

cici37 commented:
/retest

cici37 commented:
/retest

cici37 commented:
/test pull-kubernetes-bazel-build 
/test pull-kubernetes-integration

cici37 commented:
/test pull-kubernetes-node-e2e

cici37 commented:
/test pull-kubernetes-node-e2e

cici37 commented:
@liggitt @cheftako This PR is ready to review. Please have a look when have time😊

liggitt commented on pkg/controller/apis/config/v1alpha1/doc.go:
```diff
@@ -17,6 +17,12 @@ limitations under the License.
 // +k8s:deepcopy-gen=package
 // +k8s:conversion-gen=k8s.io/kubernetes/pkg/controller/apis/config
 // +k8s:conversion-gen=k8s.io/component-base/config/v1alpha1
+// +k8s:conversion-gen=k8s.io/cloud-provider/app/apis/config
```
drop the following conversion-gen references in a follow-up, we only need the external packages:
* `k8s.io/cloud-provider/app/apis/config`
* `k8s.io/cloud-provider/service/config`
* `k8s.io/controller-manager/config`

liggitt commented:
/lgtm
/approve

one note for posterity, and a few requests for follow-up fixup

cici37 commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

andrewsykim commented on pkg/controller/apis/config/types.go:
```diff
@@ -18,7 +18,9 @@ package config
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	componentbaseconfig "k8s.io/component-base/config"
+	cpconfig "k8s.io/cloud-provider/app/apis/config"
```
+1 to just `k8s.io/cloud-provider/config`

cici37 commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

cici37 commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

cici37 commented on staging/src/k8s.io/csi-translation-lib/go.mod:
```diff
@@ -15,8 +15,10 @@ require (
 	k8s.io/apimachinery => ../apimachinery
+	k8s.io/apiserver => ../apiserver
 	k8s.io/client-go => ../client-go
 	k8s.io/cloud-provider => ../cloud-provider
```
https://github.com/kubernetes/kubernetes/issues/95530
